File: gosdk-develop/accountSignature.go
package milon

import (
	"encoding/binary"
	"fmt"
	"github.com/milon-labs/milon-go-sdk/api"
	"github.com/milon-labs/milon-go-sdk/crypto"
	"github.com/milon-labs/milon-go-sdk/postcard"
	"github.com/milon-labs/milon-go-sdk/types"
)

const AuthPayerBit = 63

// AccountSignatureMode 签名模式接口
type AccountSignatureMode interface {
	isAccountSignatureMode()
}

// PubKeySignatureMode 单公钥签名模式
type PubKeySignatureMode struct {
	PublicKey crypto.PublicKey
}

func (PubKeySignatureMode) isAccountSignatureMode() {}

// MultisigKeySignatureMode 多签密钥模式
type MultisigKeySignatureMode struct {
	Index     uint8
	PublicKey crypto.PublicKey
}

func (MultisigKeySignatureMode) isAccountSignatureMode() {}

func NewAccountSignatureWithPubKey(signature crypto.Signature, pubKey crypto.PublicKey) AccountSignature {
	return AccountSignature{
		AuthBit:    types.NewBitmap64(0),
		SigBit:     types.NewBitmap64(0),
		Signatures: []crypto.Signature{signature},
		PubKey:     &pubKey,
	}
}

func NewAccountSignature(index uint8, signature crypto.Signature) (AccountSignature, error) {
	if index >= 64 {
		return AccountSignature{}, fmt.Errorf("index %d out of range (max 63)", index)
	}

	return AccountSignature{
		AuthBit:    types.NewBitmap64(0),
		SigBit:     types.NewBitmap64(1 << index),
		Signatures: []crypto.Signature{signature},
		PubKey:     nil,
	}, nil
}

type AccountSignature struct {
	AuthBit    types.Bitmap64     //1 2 4 8 ...
	SigBit     types.Bitmap64     //0=单签  非0=多签，最多64个1，根据AccountSignatureMode来获取
	Signatures []crypto.Signature //最多64个
	PubKey     *crypto.PublicKey  //PubKey为nil则SigBit为0。PubKey为有值则SigBit不为0
}

// Add 新增多签 key 的签名
func (as *AccountSignature) Add(index uint8, signature crypto.Signature) error {
	return as.AddMultisigKey(index, signature)
}

// AddMultisigKey 追加同一 auth_bit、同一授权消息下的多签 key 签名
func (as *AccountSignature) AddMultisigKey(keyIndex uint8, signature crypto.Signature) error {
	if keyIndex >= 64 {
		return fmt.Errorf("key index %d out of range (max 63)", keyIndex)
	}
	if as.PubKey != nil {
		return fmt.Errorf("pubkey mode cannot add multisig keys")
	}

	as.SigBit = as.SigBit.Set(keyIndex)
	as.Signatures = append(as.Signatures, signature)
	return nil
}

// AuthorizesIx 检查是否授权了指定的 ix
func (as *AccountSignature) AuthorizesIx(ix uint8) bool {
	return as.AuthBit.Test(ix)
}

// AuthorizesPayer 检查是否授权了 payer
func (as *AccountSignature) AuthorizesPayer() bool {
	return as.AuthBit.Test(AuthPayerBit)
}

// IxHashItem ix 哈希项
type IxHashItem struct {
	Index uint8
	Hash  api.TxHash
}

// AuthMessage 组装账户授权消息 	Blake3(MILON_ROOT || TX_AUTH_DOMAIN || chain_id || owner || auth_bit || tx_hash || ixHashes)
func (as *AccountSignature) AuthMessage(owner crypto.Address, txHash api.TxHash, ixHashes []IxHashItem) (api.TxHash, error) {
	hasher := crypto.Hash32Hasher([]byte(crypto.MilonTxAuthDomainContext))

	chainIDBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(chainIDBytes, GetChainId())
	hasher.Write(chainIDBytes)

	ownerBytes := owner.AsBytes()
	hasher.Write(ownerBytes[:])

	authBitBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(authBitBytes, uint64(as.AuthBit))
	hasher.Write(authBitBytes)

	// v3 协议：总是包含 tx_hash
	hasher.Write(txHash[:])

	for _, item := range ixHashes {
		if !as.AuthBit.Test(item.Index) {
			return api.TxHash{}, fmt.Errorf("ix index %d is not authorized in auth_bit", item.Index)
		}
		hasher.Write(item.Hash[:])
	}

	var result api.TxHash
	hasher.Sum(result[:0])
	return result, nil
}

// AuthMessageForTx 按交易上下文组装 AccountAuthMessage（验签用）
func (as *AccountSignature) AuthMessageForTx(owner crypto.Address, txHash api.TxHash, ixHashes []api.TxHash) (api.TxHash, error) {
	ixPart := CollectIxHashes(as.AuthBit, ixHashes)
	return as.AuthMessage(owner, txHash, ixPart)
}

// IsVoteGateOnly 检查是否只设置了 ix 授权位（bit0-62），但不包含 payer 位（bit63），且没有实际的签名。这用于 vote gate 场景
func (as *AccountSignature) IsVoteGateOnly() bool {
	return as.PubKey == nil &&
		len(as.Signatures) == 0 &&
		as.SigBit.Raw() == 0 &&
		(as.AuthBit.Raw()&((uint64(1)<<AuthPayerBit)-1)) != 0
	/*
		(uint64(1)<<AuthPayerBit)-1) 	==>		bit0 到 bit62 全是 1，bit63 是 0

		as.AuthBit.Raw() & ...
			AuthBit:     1xxx...xxxx  (任意值)
			掩码:        0111...1111  (只保留低 63 位)
			===========
			结果:        0xxx...xxxx  (bit63 被清零)
	*/
}

// MarshalPostcard 实现 postcard.Marshaler 接口
func (as *AccountSignature) MarshalPostcard(serializer *postcard.Serializer) error {
	err := serializer.SerializeU64(as.AuthBit.Raw())
	if err != nil {
		return fmt.Errorf("failed to serialize AuthBit: %w", err)
	}

	err = serializer.SerializeU64(as.SigBit.Raw())
	if err != nil {
		return fmt.Errorf("failed to serialize SigBit: %w", err)
	}

	err = postcard.SerializeSeq(serializer, as.Signatures, func(s *postcard.Serializer, sig crypto.Signature) error {
		return sig.MarshalPostcard(s)
	})
	if err != nil {
		return fmt.Errorf("failed to serialize TxSigs: %w", err)
	}

	err = postcard.SerializeOption(serializer, as.PubKey, func(s *postcard.Serializer, pk crypto.PublicKey) error {
		return pk.MarshalPostcard(s)
	})
	if err != nil {
		return fmt.Errorf("failed to serialize PubKey: %w", err)
	}

	return nil
}

// UnmarshalPostcard 实现 postcard.Unmarshaler 接口
func (as *AccountSignature) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	authBit, err := deserializer.DeserializeU64()
	if err != nil {
		return fmt.Errorf("failed to deserialize AuthBit: %w", err)
	}
	as.AuthBit = types.NewBitmap64(authBit)

	sigBit, err := deserializer.DeserializeU64()
	if err != nil {
		return fmt.Errorf("failed to deserialize SigBit: %w", err)
	}
	as.SigBit = types.NewBitmap64(sigBit)

	signatures, err := postcard.DeserializeSeq(deserializer, func(d *postcard.Deserializer) (crypto.Signature, error) {
		var sig crypto.Signature
		if err = sig.UnmarshalPostcard(d); err != nil {
			return sig, err
		}
		return sig, nil
	})
	if err != nil {
		return fmt.Errorf("failed to deserialize TxSigs: %w", err)
	}
	as.Signatures = signatures

	pubKey, err := postcard.DeserializeOption(deserializer, func(d *postcard.Deserializer) (crypto.PublicKey, error) {
		var pk crypto.PublicKey
		if err = pk.UnmarshalPostcard(d); err != nil {
			return pk, err
		}
		return pk, nil
	})
	if err != nil {
		return fmt.Errorf("failed to deserialize PubKey: %w", err)
	}
	as.PubKey = pubKey

	return nil
}

// AuthIx 创建 ix 授权位图（最大ix为62）
func AuthIx(ix uint8) (types.Bitmap64, error) {
	if ix >= AuthPayerBit {
		return types.NewBitmap64(0), fmt.Errorf("ix index %d out of range (max %d)", ix, AuthPayerBit-1)
	}
	return types.NewBitmap64(uint64(1) << ix), nil
}

// AuthIxes 创建多个 ix 的授权位图（最大ix为62）
func AuthIxes(indices []uint8) (types.Bitmap64, error) {
	var raw uint64
	for _, ix := range indices {
		if ix >= AuthPayerBit {
			return types.NewBitmap64(0), fmt.Errorf("ix index %d out of range (max %d)", ix, AuthPayerBit-1)
		}
		raw |= uint64(1) << ix
	}
	return types.NewBitmap64(raw), nil
}

// AuthPayer 创建 payer 授权位图
func AuthPayer() types.Bitmap64 {
	return types.NewBitmap64(uint64(1) << AuthPayerBit)
}

// AuthIxAndPayer 创建 ix + payer 联合授权位图
func AuthIxAndPayer(ix uint8) (types.Bitmap64, error) {
	if ix >= AuthPayerBit {
		return types.NewBitmap64(0), fmt.Errorf("ix index %d out of range (max %d)", ix, AuthPayerBit-1)
	}
	return types.NewBitmap64(uint64(1<<ix) | (1 << AuthPayerBit)), nil
}

// Unsigned 创建未签名的 AccountSignature（仅设置 auth_bit）
func Unsigned(authBit types.Bitmap64) AccountSignature {
	return AccountSignature{
		AuthBit:    authBit,
		SigBit:     types.NewBitmap64(0),
		Signatures: []crypto.Signature{},
		PubKey:     nil,
	}
}

// Sign 给定 auth 上下文计算授权消息并签名
func Sign(owner crypto.Address, sk crypto.SecretKeyer, authBit types.Bitmap64, txHash api.TxHash, ixHashes []IxHashItem, mode AccountSignatureMode) (*AccountSignature, error) {
	var publicKey crypto.PublicKey
	var sigBit types.Bitmap64
	var pubKeyField *crypto.PublicKey

	switch m := mode.(type) {
	case PubKeySignatureMode:
		pkAddr, err := crypto.NewAddressFromPublicKey(&m.PublicKey)
		if err != nil {
			return nil, fmt.Errorf("failed to get address from public key: %w", err)
		}
		if *pkAddr != owner {
			return nil, fmt.Errorf("public key does not match owner address")
		}
		publicKey = m.PublicKey
		sigBit = types.NewBitmap64(0)
		pubKeyField = &m.PublicKey
	case MultisigKeySignatureMode:
		if m.Index >= 64 {
			return nil, fmt.Errorf("multisig key index %d out of range (max 63)", m.Index)
		}
		publicKey = m.PublicKey
		sigBit = types.NewBitmap64(uint64(1) << m.Index)
		pubKeyField = nil
	default:
		return nil, fmt.Errorf("invalid signature mode")
	}

	accountSignature := AccountSignature{
		AuthBit:    authBit,
		SigBit:     sigBit,
		Signatures: []crypto.Signature{},
		PubKey:     pubKeyField,
	}

	authHash, err := accountSignature.AuthMessage(owner, txHash, ixHashes)
	if err != nil {
		return nil, fmt.Errorf("failed to compute auth message: %w", err)
	}

	signature, err := sk.SignFor(publicKey, authHash[:])
	if err != nil {
		return nil, fmt.Errorf("failed to sign message: %w", err)
	}

	return &AccountSignature{
		AuthBit:    authBit,
		SigBit:     sigBit,
		Signatures: []crypto.Signature{*signature},
		PubKey:     pubKeyField,
	}, nil
}

// SimulateSign 给定 auth 上下文计算授权消息但不签名
func SimulateSign(owner crypto.Address, authBit types.Bitmap64, mode AccountSignatureMode) (*AccountSignature, error) {
	var sigBit types.Bitmap64
	var pubKeyField *crypto.PublicKey

	switch m := mode.(type) {
	case PubKeySignatureMode:
		pkAddr, err := crypto.NewAddressFromPublicKey(&m.PublicKey)
		if err != nil {
			return nil, fmt.Errorf("failed to get address from public key: %w", err)
		}
		if *pkAddr != owner {
			return nil, fmt.Errorf("public key does not match owner address")
		}
		sigBit = types.NewBitmap64(0)
		pubKeyField = &m.PublicKey
	case MultisigKeySignatureMode:
		if m.Index >= 64 {
			return nil, fmt.Errorf("multisig key index %d out of range (max 63)", m.Index)
		}
		sigBit = types.NewBitmap64(uint64(1) << m.Index)
		pubKeyField = nil
	default:
		return nil, fmt.Errorf("invalid signature mode")
	}

	return &AccountSignature{
		AuthBit:    authBit,
		SigBit:     sigBit,
		Signatures: []crypto.Signature{},
		PubKey:     pubKeyField,
	}, nil
}

// CollectIxHashes 收集 auth_bit 中 ix 位对应的 IxHash
func CollectIxHashes(authBit types.Bitmap64, ixHashes []api.TxHash) []IxHashItem {
	var out []IxHashItem
	for i := uint8(0); i < 64; i++ {
		if !authBit.Test(i) {
			continue
		}
		if i == AuthPayerBit {
			continue
		}
		if int(i) < len(ixHashes) {
			out = append(out, IxHashItem{Index: i, Hash: ixHashes[i]})
		}
	}
	return out
}

File: gosdk-develop/python/milon_sdk/lib/account_signature.py
"""账户签名与鉴权（对应 Go lib/accountSignature.go）。

AuthBit 布局（64 位）：bit0-61 = 指令授权，bit62 = 保留，bit63 = gas(payer)。
AuthMessage = BLAKE3(ROOT || "milon.tx.auth.v1" || chain_id(BE u64) || owner(20B)
                     || auth_bit(LE u64) || tx_hash || 按序的已授权 ix 哈希)
"""
from __future__ import annotations

import struct
from dataclasses import dataclass, field
from typing import Optional, Protocol

from ..api.base import TxHash
from ..crypto.address import Address
from ..crypto.errors import MilonCryptoError
from ..crypto.hashes import TX_AUTH_DOMAIN, hasher
from ..crypto.keys import PublicKey, PublicKeyType
from ..crypto.secretkey import SecretKeyer
from ..crypto.signature import (
    SIGNATURE_BLS12381_SIZE,
    SIGNATURE_ED25519_SIZE,
    SIGNATURE_FNDSA512_SIZE,
    SIGNATURE_SECP256K1_SIZE,
    Signature,
    SignatureType,
)
from ..postcard import Deserializer, Serializer
from ..types import Bitmap64
from .chain import get_chain_id

AUTH_PAYER_BIT = 63
AUTH_RESERVED_BIT = 62


# ================================================================
# 签名模式
# ================================================================


class AccountSignatureMode(Protocol):
    def is_account_signature_mode(self) -> None: ...


@dataclass
class PubKeySignatureMode:
    """单公钥签名。SkipPubKey=True 时线格式省略公钥，SigBit 需从链上 signers 解析。"""

    public_key: PublicKey
    skip_pub_key: bool = False
    sig_bit: Bitmap64 = field(default_factory=lambda: Bitmap64(0))

    def is_account_signature_mode(self) -> None:
        pass


@dataclass
class MultisigKeySignatureMode:
    """多签参与者：公钥由 SigBit = 1<<Index 在链上定位。"""

    index: int
    public_key: PublicKey

    def is_account_signature_mode(self) -> None:
        pass


# ================================================================
# 账户签名
# ================================================================


class AccountSignature:
    __slots__ = ("auth_bit", "sig_bit", "signatures", "pub_key")

    def __init__(
        self,
        auth_bit: Bitmap64,
        sig_bit: Bitmap64 = None,
        signatures: list[Signature] | None = None,
        pub_key: PublicKey | None = None,
    ):
        self.auth_bit = auth_bit
        self.sig_bit = sig_bit if sig_bit is not None else Bitmap64(0)
        self.signatures = list(signatures or [])
        self.pub_key = pub_key

    # ---------------------------------------------------------- 查询
    def authorizes_ix(self, ix: int) -> bool:
        return self.auth_bit.test(ix)

    def authorizes_payer(self) -> bool:
        return self.auth_bit.test(AUTH_PAYER_BIT)

    def add_multisig_key(self, key_index: int, signature: Signature) -> None:
        if key_index >= 64:
            raise MilonCryptoError(f"key index {key_index} out of range (max 63)")
        if self.pub_key is not None:
            raise MilonCryptoError("pubkey mode cannot add multisig keys")
        self.sig_bit = self.sig_bit.set(key_index)
        self.signatures.append(signature)

    def is_vote_gate_only(self) -> bool:
        return (
            self.pub_key is None
            and len(self.signatures) == 0
            and self.sig_bit.raw() == 0
            and (self.auth_bit.raw() & ((1 << AUTH_RESERVED_BIT) - 1)) != 0
        )

    # ---------------------------------------------------------- 鉴权消息
    def auth_message(self, account: Address, tx_hash: TxHash, ix_hashes: list["IxHashItem"]) -> TxHash:
        h = hasher(TX_AUTH_DOMAIN)
        h.update(struct.pack(">Q", get_chain_id()))  # BigEndian
        h.update(account.as_bytes())
        h.update(struct.pack("<Q", self.auth_bit.raw()))  # LittleEndian（与 TxHash/IxHash 不同！）
        h.update(tx_hash.as_bytes())
        for item in ix_hashes:
            if not self.auth_bit.test(item.index):
                raise MilonCryptoError(f"ix index {item.index} is not authorized in auth_bit")
            h.update(item.hash.as_bytes())
        return TxHash(h.digest())

    # ---------------------------------------------------------- 线编解码
    def marshal_postcard(self, serializer: Serializer) -> None:
        serializer.serialize_u64(self.auth_bit.raw())
        serializer.serialize_u64(self.sig_bit.raw())
        serializer.serialize_seq(self.signatures, lambda s, sig: sig.marshal_postcard(s))
        serializer.serialize_option(
            self.pub_key is not None,
            lambda s: self.pub_key.marshal_postcard(s),
        )

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "AccountSignature":
        auth_bit = Bitmap64(d.deserialize_u64())
        sig_bit = Bitmap64(d.deserialize_u64())
        signatures = d.deserialize_seq(lambda dd: Signature.unmarshal_postcard(dd))
        pub_key = d.deserialize_option(lambda dd: PublicKey.unmarshal_postcard(dd))
        return cls(auth_bit, sig_bit, signatures, pub_key)


@dataclass
class IxHashItem:
    index: int
    hash: TxHash


# ================================================================
# AuthBit 构造辅助
# ================================================================


def auth_ix(ix: int) -> Bitmap64:
    if ix >= AUTH_RESERVED_BIT:
        raise MilonCryptoError(f"ix index {ix} out of range (max {AUTH_RESERVED_BIT - 1})")
    return Bitmap64(1 << ix)


def auth_ixes(indices: list[int]) -> Bitmap64:
    raw = 0
    for ix in indices:
        if ix >= AUTH_RESERVED_BIT:
            raise MilonCryptoError(f"ix index {ix} out of range (max {AUTH_RESERVED_BIT - 1})")
        raw |= 1 << ix
    return Bitmap64(raw)


def auth_payer() -> Bitmap64:
    return Bitmap64(1 << AUTH_PAYER_BIT)


def auth_ix_and_payer(ix: int) -> Bitmap64:
    if ix >= AUTH_RESERVED_BIT:
        raise MilonCryptoError(f"ix index {ix} out of range (max {AUTH_RESERVED_BIT - 1})")
    return Bitmap64((1 << ix) | (1 << AUTH_PAYER_BIT))


def authorizes_ix(sig: "AccountSignature", ix: int) -> bool:
    """谓词：签名是否授权了指令 ix（供 validate_wire 使用）。"""
    return sig.authorizes_ix(ix)


def authorizes_payer(sig: "AccountSignature") -> bool:
    """谓词：签名是否授权了 gas(payer) 位（供 validate_wire 使用）。"""
    return sig.authorizes_payer()


def unsigned(auth_bit: Bitmap64) -> AccountSignature:
    return AccountSignature(auth_bit=auth_bit, sig_bit=Bitmap64(0), signatures=[], pub_key=None)


# ================================================================
# 签名流程
# ================================================================


def _resolve_mode(account: Address, mode: AccountSignatureMode) -> tuple[PublicKey, Bitmap64, Optional[PublicKey]]:
    """校验签名模式与 owner 匹配，推导 (签名公钥, sigBit, PubKey 字段)。"""
    if isinstance(mode, PubKeySignatureMode):
        pk_addr = Address.from_public_key(mode.public_key)
        if pk_addr != account:
            raise MilonCryptoError("public key does not match owner address")
        if mode.skip_pub_key:
            if mode.sig_bit.raw() == 0:
                raise MilonCryptoError("SkipPubKey requires SigBit resolved from the on-chain signers list")
            return mode.public_key, mode.sig_bit, None
        return mode.public_key, Bitmap64(0), mode.public_key
    if isinstance(mode, MultisigKeySignatureMode):
        if mode.index >= 64:
            raise MilonCryptoError(f"multisig key index {mode.index} out of range (max 63)")
        return mode.public_key, Bitmap64(1 << mode.index), None
    raise MilonCryptoError("invalid signature mode")


def sign(
    account: Address,
    sk: SecretKeyer,
    auth_bit: Bitmap64,
    tx_hash: TxHash,
    ix_hashes: list[IxHashItem],
    mode: AccountSignatureMode,
) -> AccountSignature:
    public_key, sig_bit, pub_key_field = _resolve_mode(account, mode)
    acc_sig = AccountSignature(auth_bit=auth_bit, sig_bit=sig_bit, pub_key=pub_key_field)
    auth_hash = acc_sig.auth_message(account, tx_hash, ix_hashes)
    signature = sk.sign_for(public_key, auth_hash.as_bytes())
    acc_sig.signatures = [signature]
    return acc_sig


def simulate_sign(account: Address, auth_bit: Bitmap64, mode: AccountSignatureMode) -> AccountSignature:
    """无真实签名的占位签名：签名长度与真实一致，便于 gas 模拟。"""
    public_key, sig_bit, pub_key_field = _resolve_mode(account, mode)
    return AccountSignature(
        auth_bit=auth_bit,
        sig_bit=sig_bit,
        signatures=[_placeholder_signature(public_key)],
        pub_key=pub_key_field,
    )


def _placeholder_signature(pk: PublicKey) -> Signature:
    size = {
        PublicKeyType.SECP256K1: SIGNATURE_SECP256K1_SIZE,
        PublicKeyType.ED25519: SIGNATURE_ED25519_SIZE,
        PublicKeyType.BLS12381: SIGNATURE_BLS12381_SIZE,
        PublicKeyType.FNDSA512: SIGNATURE_FNDSA512_SIZE,
    }[pk.variant]
    return Signature(SignatureType(pk.variant), b"\x00" * size)


def collect_ix_hashes(auth_bit: Bitmap64, ix_hashes: list[TxHash]) -> list[IxHashItem]:
    out: list[IxHashItem] = []
    for i in range(AUTH_RESERVED_BIT):
        if not auth_bit.test(i):
            continue
        if i < len(ix_hashes):
            out.append(IxHashItem(index=i, hash=ix_hashes[i]))
    return out
